### ISSUE TITLE
18.04 snap - can't find input file in /tmp

### ISSUE THREAD

**Issue by wrighrc** [smoser/pdftk#1]
Hi can you tell me if this should work?    Any suggestions?    Thanks.

charles@x1:/tmp$ pdftk -version
pdftk 2.02 a Handy Tool for Manipulating PDF Documents
Copyright (c) 2003-13 Steward and Lee, LLC - Please Visit: www.pdftk.com
This is free software; see the source code for copying conditions. There is
NO warranty, not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

charles@x1:/tmp$ snap list | grep pdftk
pdftk                2.02-4     1     stable    smoser     -

charles@x1:/tmp$ ls -l in.pdf 
-rw-r--r-- 1 charles charles 9639 Jul 13 08:15 in.pdf

charles@x1:/tmp$ pdftk PROMPT
Please enter a filename for an input PDF:
in.pdf
Error: Unable to find file.
Error: Failed to open PDF file: 
   in.pdf
Done.  Input errors, so no output created.

charles@x1:/tmp$ pdftk /tmp/in.pdf input_pw 'redacted' output /tmp/out.pdf
Error: Unable to find file.
Error: Failed to open PDF file: 
   /tmp/in.pdf
Errors encountered.  No output created.
Done.  Input errors, so no output created.

charles@x1:/tmp$ pdftk PROMPT
Please enter a filename for an input PDF:
/tmp/in.pdf
Error: Unable to find file.
Error: Failed to open PDF file: 
   /tmp/in.pdf
Done.  Input errors, so no output created.

**Comment by smoser** [comment]
Hi,
 I banged my head against a wall once trying to figure out why the vlc snap couldn't play a video for me.  You seem to have some of the same behaviors as I do, putting files in /tmp. 

Snaps have a different view of the filesystem when confined.  So /tmp that you see is different than the confined 'pdftk' sees.  So when pdftk tried to open the file '/tmp/in.pdf' it was not there.

The pdftk snap has the "home" [plug](https://docs.snapcraft.io/reference/interfaces) so it can read files in  your `$HOME`.

The easiest workaround then is simple: do not use `/tmp`, use `$HOME`.

There might be other tricks with bash, something like:

     pdftk <(cat in.pdf) input_pw 'redacted' output >(cat >outfile)

Access to stdin and stdout in a snap can sometimes be weird too though ([bug 1766667](https://bugs.launchpad.net/snappy/+bug/1766667)).

Hope this helps.

**Comment by smoser** [comment]
closing. re-open if you think it needs to be. there really isnt much i can do.

**Comment by masgo** [comment]
Hi @smoser 
I have a similar problem, but one where you might be able to help. I too want to use pdftk on files which are not in my home directory, but on a network share. 

From what I learned so far about snaps is, that there is the `removable-media` interface. This would allow the snap to access `/media`. My network share is mounted there, so offering `removable-media` as a "plug" would allow pdftk to access files on the network share. 

It would also drastically widen the possibilities, since one could mount all sorts of things there which would not be possible or practical in the home directory.

**can you add the  `removable-media` plug?**

**Comment by smoser** [comment]
@masgo, I added the 'removable-media' plug, and released that snap to stable channel today.

**Comment by mh11-mh11** [comment]
Hi all,

I'm having a similar issue, not sure, whether the 'removable-media' plug already solved this (or if it's a snap issue maybe?)?

My files are on a NFS share and I get also the ''unable to find file" error.
Of course, I worked around this by copying the files to my /home folder, but longterm, it would be great to use input files from everywhere.

Could you please check this or let me know if this is snap's or my issue?
Thx!

**Comment by smoser** [comment]
@mh11-mh11 ,

Hi, I'm sorry. I honestly didn't test this  myself until just now. It looks to me like you  may have to explicitly enable access to /media after install.

Can you try:

    snap connect pdftk:removable-media

And see if then you can use /media?  And please report back.

If that's not it, then I'm out of ideas and we would need to ask in snap forum.

**Comment by mh11-mh11** [comment]
Hi sorry, for the late response, did not see the notification :-(

It did not work for me as I'm trying to access a NFS share, but I believe, it's a good staring point, I just need to find the name for the correct connector. Will keep you posted!

Thanks for the hint!

**Comment by mh11-mh11** [comment]
oh, I just read your post again (didn't see this the first time). You're writing "enable access to /media ".
Well, my NFS shares are mounted under /mnt (as I mount them via /etc/fstab), maybe that's the point?

How can I enable access to /mnt?

BTW: I have the exact same problem w/ GIMP, which I just installed, so it looks more like a snap issue than a pdftk issue, anyways, any help is very appreciated. Thanks!

ok, played a bit more around, it actually looks like a pdftk issue, sorry to say.
For comprarison: I could solve my (identical) GIMP issue actualy w/
`connect gimp:removable-media`
so, this means, "removable-media" gives me access to /mnt.

However, this does not work for pdftk, even a disconnect followed by a reconnect does not help.

Here's what snap connections shows (extract):

`removable-media           gimp:removable-media                   :removable-media                 manual

removable-media           pdftk:removable-media                  :removable-media                 manual`

Anyway, thank you very much again, at least you could solve my GIMP issue :-)

**Comment by smoser** [comment]
I'm  not sure that you can give access to /mnt directly
Possibly you could bind-mount /mnt over /media/some-path.
  mount --rbind /mnt /media/nfs

**Comment by ntc2** [comment]
I just ran into this and I think it should be reopened. The pdftk used to handle files in /tmp just fine, and now it mysteriously fails, with no indication of what the problem is (it says "file not found", but the file exists; there's no indication that the problem is due to pdftk being packaged in a snap).

I don't know anything about snap packages, but my impression from above is that this is not easy to fix, i.e. there is no "plug" (?) for getting access to /tmp? So maybe it's hard to actually fix the problem, but perhaps the problematic situation could be detected and a better error message given, or the current "missing file" error message could be expanded to mention that pdftk is broken for input files in /tmp?

**Comment by JazzTp** [comment]
@ntc2

**pdftk 2.02** (on Linux Ubuntu 18.04.4) apparently fails to read _and write_ from/to any folder under /tmp (tried granting all permissions to user group and others and using a different mktemp pattern with no dot).

~(Would anybody please know of any fixed build available?)~
Adopting a mv workaround (of course some day a fixed build will make it upstream so all Linux distributions get it and everybody will be happier :smile:)

**Comment by smoser** [comment]
There really isn't anything that a snap can do here.  You just have to not use /tmp.

https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1851250
https://askubuntu.com/questions/1188811/snap-how-to-grant-tmp-access-to-snap-chromium

**Comment by JazzTp** [comment]
> There really isn't anything that a snap can do here. You just have to not use /tmp.
> 
> https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1851250
> https://askubuntu.com/questions/1188811/snap-how-to-grant-tmp-access-to-snap-chromium

Yes, done. Thank you for the interesting links :+1:

**Comment by flintiii** [comment]
TADA...

A possible silver bullet is to reboot the system (what a terrible solution!!!)

Flint

**Comment by m0ritz3** [comment]
If it would be possible to install pdftk with classic confinement every problem discussed here would be solved. I have got a second hard drive mounted at /datadisk and thus cannot get any snap with strict confinement access my files. 
I think this is against Linux philosophy. I want to make any changes I want to my system even if I end up damaging it. Strict confinement as the default is a perfectly reasonable choice though. 

TLDR: Please offer classic confinement.

**Comment by ifokkema** [comment]
@m0ritz3, have you tried this workaround?
https://github.com/smoser/pdftk/issues/7#issuecomment-479782445

It works great for me!

**Comment by m0ritz3** [comment]
@ifokkema I ditched snap completely and changed my OS to Linux Mint 20 Ulyana. So I cannot confirm whether this workaround would have worked for me.

**Comment by ntc2** [comment]
> I just ran into this and I think it should be reopened. The pdftk used to handle files in /tmp just fine, and now it mysteriously fails, with no indication of what the problem is (it says "file not found", but the file exists; there's no indication that the problem is due to pdftk being packaged in a snap).
> 
> I don't know anything about snap packages, but my impression from above is that this is not easy to fix, i.e. there is no "plug" (?) for getting access to /tmp? So maybe it's hard to actually fix the problem, but perhaps the problematic situation could be detected and a better error message given, or the current "missing file" error message could be expanded to mention that pdftk is broken for input files in /tmp?

Just ran into this again 3.5 years later, completely forgetting I had this problem before. Took me an hour to find this issue and see that I've already commented on it in the past :facepalm:
